doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+    -   Convert subsources to `CREATE TABLE` statements with the same name
+    -   Change the `CREATE SOURCE` statement to point to the current progress collection and
+        change the source name to `<source>_progress`
+    -   Drop the existing `CREATE SUBSOURCE <source>_progress` statement since this will no
+        longer have a collection to own
+    -   Drop the existing top-level collection tied to the source, since this is already
+        unused and will no longer be owned by any statement
+
+#### Exposing available upstream references to users
+
+Since creating a source will no longer automatically create any subsources, users that
+do not already know all the upstream references available in their system may find it difficult
+to know which `CREATE TABLE .. FROM SOURCE` statements they could use to ingest their
+upstream data. Additionally, the Web Console will need a way to present a dynamic UI
+to users after they have created a source to allow them to create any downstream tables
+from a selection of the available upstream references of that source.
+
+We have two options available to us for exposing this information to users:
+
+1.  Add a new statement to fetch from the upstream all possible 'upstream references' available
+    for ingestion.
+2.  Create a new system table that periodically updates with the set of available upstream
+    references in each sources, and a new statement to trigger a refresh of this system table.
+
+The 2nd option is more complex to implement, but presents the information in a form that
+is more similar to other similar information in Materialize. However since the data can
+become out of date, it's important that the refresh statement is highlighted prominently
+in user-facing documentation and related guides.
+
+The proposal for the 2nd option (a new system table):
+
+Introduce a `mz_catalog.mz_available_source_references` with the following columns:
+
+-   `source_id`: The source (e.g. a Postgres, MySQL, Kafka source) this reference is available for.
+-   `reference`: The 'external reference' for this available object, such that using `reference`
+    in a `CREATE TABLE <table_name> FROM SOURCE <source_id> (REFERENCE <reference>)` would
+    construct a table ingesting this upstream reference. This is essentially the fully-qualified
+    upstream name of the reference.
+-   `columns`: If available, an array of of columns that this external reference would have if it
+    were ingested as a table in Materialize. This will be NULL for source types where an
+    external registry is needed to determine the upstream schema (e.g. Kafka).
+-   `name`: The object name of the reference. E.g. for a Postgres source this is the table name,
+    and for kafka is the topic name.
+-   `namespace`: Any namespace that this reference object exists in, if applicable. For Postgres
+    this would be `<database>.<schema>`, for MySQL this would be `<schema>`, and for Kafka this
+    would be NULL.
```

Comment:
This seems semi redundant with the  `reference` column?

Reply:
I imagine the redundancy being helpful if we want to also expose the available schemas for PostgreSQL and MySQL sources, if users want to create tables for all objects in the upstream namespace. Otherwise, we'd have to parse that out of `name` in the Console logic, no?

Reply:
Yeah, sorry, I meant it the opposite way around—I think we should ditch `reference` and have only `name`/`namespace`/`super_namespace`. I think the console will basically always want to construct names as `quoteIdent(name1) + "." + quoteIdent(name2) + "." + quoteIdent(name3)`.